src/bitcoin-cli.cpp

```diff
@@ -757,8 +762,9 @@ static UniValue ConnectAndCallRPC(BaseRequestHandler* rh, const std::string& str
             }
             break; // Connection succeeded, no need to retry.
         } catch (const CConnectionFailed&) {
-            if (fWait) {
-                UninterruptibleSleep(std::chrono::milliseconds{1000});
+            const int64_t now = GetTime<std::chrono::seconds>().count();
+            if (fWait && (timeout == 0 || now < deadline)) {
```

Comment:
The error I'm seeing is a bit confusing. It would be good to raise a specific error for the timeout value exceeded.
```
$ bitcoind -regtest -daemon ; bitcoin-cli -regtest -rpcwait -rpcwaittimeout=3 -getinfo
Bitcoin Core starting
error: server in warmup
```